assets/js/common/OperationModals/ClusterMaintenanceChangeModal.jsx

```diff
@@ -0,0 +1,128 @@
+import React, { useState } from 'react';
+import { concat, flatMap, flow, get, map, uniqBy } from 'lodash';
+import Label from '@common/Label';
+import Select from '@common/Select';
+import Switch from '@common/Switch';
+import OperationModal from './OperationModal';
+
+const NOT_SELECTED = 'Select a cluster resource';
+const CLUSTER_LABEL = 'Cluster (full maintenance)';
+const NODE_MAINTENANCE_STATE = 'Maintenance';
+const CLUSTER_SCOPE = 'cluster';
+const NODE_SCOPE = 'node';
+const RESOURCE_SCOPE = 'resource';
+
+const NOT_SELECTED_OPTION = {
+  value: {
+    id: NOT_SELECTED,
+    maintenance: false,
+  },
+  key: 'not_selected',
+};
+
+const renderOption = ({ value: { id } }) => id;
+
+function ClusterMaintenanceChangeModal({
+  clusterDetails,
+  isOpen = false,
+  onRequest,
+  onCancel,
+}) {
+  const [checked, setChecked] = useState(false);
+  const [resource, setResource] = useState(NOT_SELECTED_OPTION.value);
+  const [maintenanceState, setMaintenanceState] = useState(false);
+
+  const clusterMaintenance = get(clusterDetails, 'maintenance_mode', false);
+  const clusterNodes = get(clusterDetails, 'nodes', []);
+  const stoppedResource = get(clusterDetails, 'stopped_resources', []);
+
+  const clusterOption = {
+    value: {
+      id: CLUSTER_LABEL,
+      maintenance: clusterMaintenance,
+      scope: CLUSTER_SCOPE,
+    },
+    key: 'cluster',
+  };
+
+  const nodeOptions = map(clusterNodes, ({ name, status }) => ({
+    value: {
+      id: name,
+      maintenance: status === NODE_MAINTENANCE_STATE,
+      scope: NODE_SCOPE,
+    },
+    key: name,
+  }));
+
+  // flatMap all resources and parents from all nodes
+  // and concat stopped resources
+  const resourceOptions = flow([
+    (nodes) => flatMap(nodes, ({ resources }) => resources),
+    (resources) =>
+      flatMap(resources, (res) => (res.parent ? [res.parent, res] : [res])),
+    (resources) => concat(resources, stoppedResource),
+    (resources) => uniqBy(resources, 'id'),
+    (resources) =>
+      map(resources, ({ id, managed }) => ({
+        value: { id, maintenance: !managed, scope: RESOURCE_SCOPE },
+        key: id,
+      })),
+  ])(clusterNodes);
+
+  const allOptions = concat(
+    NOT_SELECTED_OPTION,
+    clusterOption,
+    nodeOptions,
+    resourceOptions
+  );
+
+  return (
+    <OperationModal
+      title="Cluster Maintenance"
+      description="Update cluster, node or resource maintenance state"
+      operationText="Maintenance"
+      applyDisabled={!checked || resource.maintenance === maintenanceState}
+      checked={checked}
+      isOpen={isOpen}
+      onChecked={() => setChecked((prev) => !prev)}
+      onRequest={() =>
+        // send node_id or resource_id only if the option is of that type
+        onRequest({
+          maintenance: maintenanceState,
```

Comment:
Maybe moving this block to a seperate function could improve the maintability and readablity? 

```
const handleRequest = () => {
  const requestData = {
    maintenance: maintenanceState,
    ...(resource.scope === NODE_SCOPE && { node_id: resource.id }),
    ...(resource.scope === RESOURCE_SCOPE && { resource_id: resource.id }),
  };

  onRequest(requestData);
};


and just have 
<OperationModal
  ....
  onRequest={handleRequest} 
  ....
/>
```

It is more my personal preference so feel free :)

Reply:
I think it is fine as it is. It is a fairly simple thing